Android and jenkins series. Blog post #2. (#6336)
jenkins-infra/jenkins.io: content/blog/2023/05/03/2023-05-03-android-and-jenkins-releases.adoc

Therefore, let's go back to the drawing board and add the command to our Docker image.

[source,dockerfile]
----
# Install GitHub command line tool
ENV GITHUB_TOKEN $GITHUB_TOKEN
RUN curl -fsSL https://cli.github.com/packages/githubcli-archive-keyring.gpg | dd of=/usr/share/keyrings/githubcli-archive-keyring.gpg && \
    chmod go+r /usr/share/keyrings/githubcli-archive-keyring.gpg && \
    echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/githubcli-archive-keyring.gpg] https://cli.github.com/packages stable main" | tee /etc/apt/sources.list.d/github-cli.list > /dev/null && \
    apt update && apt install -y --no-install-recommends gh
----

Once that's done, we need to use link:https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/about-authentication-with-a-github-app[GitHub App authentication] to enable `gh` to use our credentials.
To do this, we have to install the link:https://plugins.jenkins.io/github-branch-source/[GitHub Branch Source plugin] and then create a link:https://www.jenkins.io/blog/2020/04/16/github-app-authentication/[GitHub Application].

> No need to replicate here the link:https://github.com/jenkinsci/github-branch-source-plugin/blob/master/docs/github-app.adoc[efficient documentation] existing on GitHub.

Reviewer: ```suggestion
No need to replicate here the link:https://github.com/jenkinsci/github-branch-source-plugin/blob/master/docs/github-app.adoc[efficient documentation] existing on GitHub.
```
I wasn't sure how to rephrase this, as I wasn't sure if the intention is to say "no need to recreate the full documentation" or "the github documentation is accurate and should be copied over"

If you could provide any insight on this, I'd appreciate it very much!

Author: Thanks Kevin. 😊 
What I wanted to say is the documentation on GitHub is spot on, a link to it should be enough. 🤷

Reviewer: Ahh okay that makes sense.

Therefore, let's go back to the drawing board and add the command to our Docker image.

[source,dockerfile]
----
# Install GitHub command line tool
ENV GITHUB_TOKEN $GITHUB_TOKEN
RUN curl -fsSL https://cli.github.com/packages/githubcli-archive-keyring.gpg | dd of=/usr/share/keyrings/githubcli-archive-keyring.gpg && \
    chmod go+r /usr/share/keyrings/githubcli-archive-keyring.gpg && \
    echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/githubcli-archive-keyring.gpg] https://cli.github.com/packages stable main" | tee /etc/apt/sources.list.d/github-cli.list > /dev/null && \
    apt update && apt install -y --no-install-recommends gh
----

Once that's done, we need to use link:https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/about-authentication-with-a-github-app[GitHub App authentication] to enable `gh` to use our credentials.
To do this, we have to install the link:https://plugins.jenkins.io/github-branch-source/[GitHub Branch Source plugin] and then create a link:https://www.jenkins.io/blog/2020/04/16/github-app-authentication/[GitHub Application].

> No need to replicate here the link:https://github.com/jenkinsci/github-branch-source-plugin/blob/master/docs/github-app.adoc[efficient documentation] existing on GitHub.

Reviewer: ```suggestion
The link:https://github.com/jenkinsci/github-branch-source-plugin/blob/master/docs/github-app.adoc[existting documentation] on GitHub is exactly what we need, so a link to this should suffice.
```
does this make a bit more sense then @gounthar? If not, I'll let you handle that one since I don't want to suggest something incorrectly.

Author: Yes, thank you Kevin. 👍

You have to link:https://developers.google.com/android-publisher/getting_started#enable[enable the Android Publisher API] for that project.

video::eXJBIkHNB48[youtube, width=839, height=473, role=center]

Then, you have to link:https://developers.google.com/android-publisher/getting_started#existing[link] your Google Play developer account to the GCP project.

video::XaokL2ku4JA[youtube, width=839, height=473, role=center]

After this, you need to link:https://cloud.google.com/iam/docs/service-accounts-create[create a service account].

video::hAHvZe1XklU[youtube, width=839, height=473, role=center]

Then create a link:https://cloud.google.com/iam/docs/keys-create-delete[key].

video::LdMSK1d63Sw[youtube, width=839, height=473, role=center]

To set up the necessary credentials for publishing our app to the Play Store, we'll need to create an environment variable in Jenkins.
To do this, we first need to install the link:https://plugins.jenkins.io/envinject/[Environment Injector plugin].
Once that's done, we can grant the necessary permissions to our service account so that it can publish the app on our behalf.

video::LXVydeeMnSU[youtube, width=839, height=473, role=center]

> And we're finally ready to publish our app thanks to gradle on Jenkins.

Reviewer: ```suggestion
And we're finally ready to publish our app thanks to Gradle on Jenkins.
```
should this instance be capitalized like the rest?

Publishing tasks
----------------
[...]
publishBundle - Uploads App Bundle for all variants.
   See https://github.com/Triple-T/gradle-play-publisher#publishing-an-app-bundle
[...]
BUILD SUCCESSFUL in 1s
1 actionable task: 1 executed
----

As we did not store the generated `jks` file in the repo, we have to use a variable to hold the value.
On your machine, it would work with something like:

[source,bash]
----
export ANDROID_PUBLISHER_CREDENTIALS=`cat *json`
----

On Jenkins, we will create a secret.

video::XkORY9nbgak[youtube, width=839, height=473, role=center]

The secret is now available under the `android-publisher-credentials` key.

The triplet link:https://github.com/Triple-T/gradle-play-publisher#common-configuration[documentation] tells us that we can set up a configuration in the build.gradle file like:

[source,groovy]
----
play {
    // Overrides defaults
    track.set("internal")
    updatePriority.set(2)
    releaseStatus.set(ReleaseStatus.DRAFT)
    // ...
}
----

Gradle Play Publisher supports uploading both the App Bundle and APK, and can promote those artifacts to different tracks.
You can customize how your artifacts are published using several options:

> * `track`: The target stage for an artifact, such as `internal`/`alpha`/`beta`/`production` or any custom track.
> ** Defaults to internal.
> * `releaseStatus`: The type of release, such as `ReleaseStatus.COMPLETED`, `ReleaseStatus.DRAFT`, `ReleaseStatus.HALTED`, or `ReleaseStatus.IN_PROGRESS`.
> ** Defaults to `ReleaseStatus.COMPLETED`.
> * `userFraction`: The percentage of users who will receive a staged release.
> ** This is only applicable where `releaseStatus=[IN_PROGRESS/HALTED]`
> ** defaults to `0.1` (10%).
> * `updatePriority`: Sets the update priority for a new release. Refer to link:https://developer.android.com/guide/playcore/in-app-updates[Google's documentation] for more information.
> ** Defaults to the API value.

Reviewer: ```suggestion
* `track`: The target stage for an artifact, such as `internal`/`alpha`/`beta`/`production` or any custom track.
** Defaults to internal
* `releaseStatus`: The type of release, such as `ReleaseStatus.COMPLETED`, `ReleaseStatus.DRAFT`, `ReleaseStatus.HALTED`, or `ReleaseStatus.IN_PROGRESS`.
** Defaults to `ReleaseStatus.COMPLETED`
* `userFraction`: The percentage of users who will receive a staged release.
** This is only applicable where `releaseStatus=[IN_PROGRESS/HALTED]`.
** defaults to `0.1` (10%)
* `updatePriority`: Sets the update priority for a new release.
Refer to link:https://developer.android.com/guide/playcore/in-app-updates[Google's documentation] for more information.
** Defaults to the API value
```
removed periods from sentence fragments in list, added where needed
